ui/components/app/currency-input/currency-input.js

```diff
@@ -41,84 +45,95 @@ export default function CurrencyInput({
 
+  const assetDecimals = asset?.decimals || NATIVE_CURRENCY_DECIMALS;
+
   const preferredCurrency = useSelector(getNativeCurrency);
   const secondaryCurrency = useSelector(getCurrentCurrency);
-  const conversionRate = useSelector(getConversionRate);
-  const showFiat = useSelector(getShouldShowFiat);
+
+  const primarySuffix =
+    asset?.symbol || preferredCurrency || EtherDenomination.ETH;
+  const secondarySuffix = secondaryCurrency.toUpperCase();
+  const isLongSymbol = (primarySuffix?.length || 0) > LARGE_SYMBOL_LENGTH;
+
+  const [shouldDisplayFiatIfAvailable, setShouldDisplayFiatIfAvailable] =
+    useState(featureSecondary);
+  const isFiatAvailable = useSelector(getShouldShowFiat);
+
+  const shouldUseFiat =
+    isFiatAvailable && Boolean(shouldDisplayFiatIfAvailable);
+  const isTokenPrimary = !shouldUseFiat;
+
+  const [tokenDecimalValue, setTokenDecimalValue] = useState('0');
+  const [fiatDecimalValue, setFiatDecimalValue] = useState('0');
+
   const chainId = useSelector(getCurrentChainId);
   const { ticker, type } = useSelector(getProviderConfig);
   const isOriginalNativeSymbol = useIsOriginalNativeTokenSymbol(
     chainId,
     ticker,
     type,
   );
-  const hideSecondary = !showFiat;
-  const primarySuffix = preferredCurrency || EtherDenomination.ETH;
-  const secondarySuffix = secondaryCurrency.toUpperCase();
 
-  const [isSwapped, setSwapped] = useState(false);
-  const [newHexValue, setNewHexValue] = useState(hexValue);
-  const [shouldDisplayFiat, setShouldDisplayFiat] = useState(featureSecondary);
-  const shouldUseFiat = hideSecondary ? false : Boolean(shouldDisplayFiat);
-
-  const getDecimalValue = () => {
-    const decimalValueString = shouldUseFiat
-      ? getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: secondaryCurrency,
-          conversionRate,
-          numberOfDecimals: 2,
-        })
-      : getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: EtherDenomination.ETH,
-          numberOfDecimals: 8,
-        });
-
-    return Number(decimalValueString) || 0;
-  };
+  const tokenToFiatConversionRate = useTokenExchangeRate(asset?.address);
 
-  const initialDecimalValue = hexValue ? getDecimalValue() : 0;
+  const processNewDecimalValue = useProcessNewDecimalValue(
+    assetDecimals,
+    isTokenPrimary,
+    tokenToFiatConversionRate,
+  );
 
   const swap = async () => {
     await onPreferenceToggle();
-    setSwapped(!isSwapped);
-    setShouldDisplayFiat(!shouldDisplayFiat);
+    setShouldDisplayFiatIfAvailable(!shouldDisplayFiatIfAvailable);
   };
 
+  // if the conversion rate is undefined, do not allow a fiat input
+  useEffect(() => {
+    if (isTokenPrimary) {
+      return;
+    }
+
+    if (!tokenToFiatConversionRate) {
+      onPreferenceToggle();
+      setShouldDisplayFiatIfAvailable(false);
+    }
+  }, [tokenToFiatConversionRate, isTokenPrimary, onPreferenceToggle]);
+
   const handleChange = (newDecimalValue) => {
```

Comment:
Found another bug:
1. Toggle to fiat
2. Type in 100 USD
3. Change token to USDC -> $100 is not converted to the correct USDC token amount. Should this be converted? OR all inputs reset to 0 on token change?

My initial guess is that handleChange (or the useEffect) is using an outdated conversion rate so it needs to be a useCallback

Reply:
Great catch – resolved.